[mlas] Speed up tanhf activation function

r-devulap opened the pull request:
### Description
New faster algorithm for `tanhf `activation function using Intel SVML. 

### Motivation and Context
Improves performance of `tanhf` by nearly 40%. The newer algorithm also fixes a bug in the current `tanhf` algorithm which goes out of bounds [-1, 1]. Example: for `x = +0x1.06417ep+003`, `tanhf= +0x1.000002p+000`. 

```
Benchmark                                                 Time             CPU      Time Old      Time New       CPU Old       CPU New
--------------------------------------------------------------------------------------------------------------------------------------
[BM_Tanh vs. BM_Tanh]/40000/real_time                  -0.3822         -0.3825         15059          9304         15035          9283
[BM_Tanh vs. BM_Tanh]/80000/real_time                  -0.3845         -0.3844         30055         18499         29998         18467
[BM_Tanh vs. BM_Tanh]/160000/real_time                 -0.3146         -0.3144         17803         12203         17762         12178
[BM_Tanh vs. BM_Tanh]/320000/real_time                 -0.3495         -0.3491         32840         21362         32724         21300
[BM_Tanh vs. BM_Tanh]/640000/real_time                 -0.3563         -0.3568         62902         40487         62754         40361
[BM_Tanh vs. BM_Tanh]/1280000/real_time                -0.3326         -0.3333        128536         85780        128102         85408
OVERALL_GEOMEAN                                        -0.3538         -0.3539             0             0             0             0

```

yufenglee commented:
/azp run ONNX Runtime Web CI Pipeline,Linux CPU CI Pipeline,Linux GPU CI Pipeline,orttraining-amd-gpu-ci-pipeline,Android CI Pipeline,iOS CI Pipeline,ONNX Runtime React Native CI Pipeline

yufenglee commented:
/azp run Linux OpenVINO CI Pipeline,Linux QNN CI Pipeline,MacOS CI Pipeline,orttraining-amd-gpu-ci-pipeline,orttraining-linux-ci-pipeline,orttraining-linux-gpu-ci-pipeline,orttraining-ortmodule-distributed,onnxruntime-binary-size-checks-ci-pipeline,Big Models,Android CI Pipeline

yihonglyu reviewed (commented):
Please add a benchmark for the tanh activation function in `onnxruntime/test/mlas/bench/`. Once you've done that, make sure to record the performance number both with and without your patch in the commit message.

r-devulap commented:
> Please add a benchmark for the tanh activation function in onnxruntime/test/mlas/bench/. 

There is already a benchmark for tanhf `BM_Tanh`. Is this not sufficient? https://github.com/microsoft/onnxruntime/blob/69cfcba38a60d65498f94cde30cb9c2030f7255b/onnxruntime/test/onnx/microbenchmark/activation.cc#L342-L344

> Once you've done that, make sure to > record the performance number both with and without your patch in the commit message.

The performance numbers of `BM_Tanh` before and after have already been included in the commit message: See https://github.com/microsoft/onnxruntime/pull/20612/commits/c6c93092f333650f126d0a83bce3340a4a179eb4

yufenglee commented:
/azp run ONNX Runtime Web CI Pipeline,Linux CPU CI Pipeline,Linux GPU CI Pipeline,orttraining-amd-gpu-ci-pipeline,Android CI Pipeline,iOS CI Pipeline,ONNX Runtime React Native CI Pipeline

yufenglee commented:
/azp run Linux OpenVINO CI Pipeline,Linux QNN CI Pipeline,MacOS CI Pipeline,orttraining-amd-gpu-ci-pipeline,orttraining-linux-ci-pipeline,orttraining-linux-gpu-ci-pipeline,orttraining-ortmodule-distributed,onnxruntime-binary-size-checks-ci-pipeline,Big Models,Android CI Pipeline

yufenglee commented on onnxruntime/core/mlas/lib/tanh.cpp:
```diff
@@ -83,69 +73,48 @@ Return Value:
+
+    size_t count = 0;
+    while (count < N) {
+        if (N - count >= 4) {
```
there needs a check for each iteration with this change. If N is large, the previous version can save a significant amount of instructions

r-devulap replied:
For large values of N, the CPU branch predictor should be able to branch predict this branch pretty easily. It will only miss at the very last iteration for the tail but when N is large, one single branch miss should hardly matter in terms of performance. It does bring the benefit of processing the entire array contained in a single loop.

yufenglee commented on onnxruntime/core/mlas/lib/amd64/TanhKernelFma3.asm:
```diff
@@ -64,39 +64,33 @@ INCLUDE TransKernelCommon.inc
-        vmulps  ymm2,ymm0,ymm2                  ; p = x * p
-        vdivps  ymm0,ymm2,ymm3                  ; tanh = p / q
+        vandps      ymm0,ymm10,YMMWORD PTR [rcx]
+	vmovaps	    ymm3, ymm5
```
>	 [](http://example.com/codeflow?start=0&length=1)

nit: please replace the tab with space

r-devulap replied:
done.

yufenglee commented on onnxruntime/core/mlas/lib/x86_64/TanhKernelFma3.S:
```diff
@@ -48,71 +49,61 @@ Return Value:
+        vmovups     ymm15,YMMWORD PTR [r10+rdx*4]
+        vmaskmovps  ymm0,ymm15,YMMWORD PTR [rdi]
+        vandps      ymm0,ymm10,ymm0
+	vmovaps	    ymm3, ymm5
```
>	 [](http://example.com/codeflow?start=0&length=1)

nit: same here

r-devulap replied:
done.

yufenglee commented:
/azp run ONNX Runtime Web CI Pipeline,Linux CPU CI Pipeline,Linux GPU CI Pipeline,orttraining-amd-gpu-ci-pipeline,Android CI Pipeline,iOS CI Pipeline,ONNX Runtime React Native CI Pipeline

yufenglee commented:
/azp run Linux OpenVINO CI Pipeline,Linux QNN CI Pipeline,MacOS CI Pipeline,orttraining-amd-gpu-ci-pipeline,orttraining-linux-ci-pipeline,orttraining-linux-gpu-ci-pipeline,orttraining-ortmodule-distributed,onnxruntime-binary-size-checks-ci-pipeline,Big Models,Android CI Pipeline

yufenglee commented:
/azp run Linux Android Emulator QNN CI Pipeline, Linux CPU Minimal Build E2E CI Pipeline, Linux GPU TensorRT CI Pipeline, Windows ARM64 QNN CI Pipeline, Windows CPU CI Pipeline , Windows GPU CI Pipeline, Windows GPU TensorRT CI Pipeline, Windows x64 QNN CI Pipeline

yufenglee commented:
You need sign the license/cla agreement to move on.

r-devulap commented:
> You need sign the license/cla agreement to move on.

Yup, working on it.

yufenglee commented:
/azp run ONNX Runtime Web CI Pipeline,Linux CPU CI Pipeline,Linux GPU CI Pipeline,orttraining-amd-gpu-ci-pipeline,Android CI Pipeline,iOS CI Pipeline,ONNX Runtime React Native CI Pipeline

yufenglee commented:
/azp run Linux OpenVINO CI Pipeline,Linux QNN CI Pipeline,MacOS CI Pipeline,orttraining-amd-gpu-ci-pipeline,orttraining-linux-ci-pipeline,orttraining-linux-gpu-ci-pipeline,orttraining-ortmodule-distributed,onnxruntime-binary-size-checks-ci-pipeline,Big Models,Android CI Pipeline

yufenglee commented:
/azp run Linux OpenVINO CI Pipeline,Linux QNN CI Pipeline,MacOS CI Pipeline,orttraining-amd-gpu-ci-pipeline,orttraining-linux-ci-pipeline,orttraining-linux-gpu-ci-pipeline,orttraining-ortmodule-distributed,onnxruntime-binary-size-checks-ci-pipeline,Big Models,Android CI Pipeline

yufenglee commented:
/azp run Linux Android Emulator QNN CI Pipeline, Linux CPU Minimal Build E2E CI Pipeline, Linux GPU TensorRT CI Pipeline, Windows ARM64 QNN CI Pipeline, Windows CPU CI Pipeline , Windows GPU CI Pipeline, Windows GPU TensorRT CI Pipeline, Windows x64 QNN CI Pipeline

r-devulap commented:
@yufenglee Couple of questions I need help with: 

1) I'm unable to replicate the `Windows CI pipeline` run fail locally (fail log [here](https://dev.azure.com/onnxruntime/onnxruntime/_build/results?buildId=1382352&view=logs&j=2c350baf-0cef-5ef1-cbeb-e5cd88c89a11&t=e2851d88-2248-5227-ac0b-d999486a4bf5)). The test `LSTMTest.BackwardCompute` passes for me. Any pointer on why its behaving different? 

2) The other failure in `Windows GPU CI Pipeline` (see [log](https://dev.azure.com/onnxruntime/onnxruntime/_build/results?buildId=1382353&view=logs&j=3032dfba-5baf-5872-0871-2e69cb7f4b6a&t=91e76d45-ff6f-5804-5ba0-eca18ac9ce48)) fails a test `onnxruntime_test_all` -> `ModelTests/ModelTest.Run/fp16_coreml_FNS_Candy_opset7_CPU`. But my local build doesn't contain the `ModelTests*` set at all. How do I build these tests?

yufenglee commented:
> @yufenglee Couple of questions I need help with:
> 
> 1. I'm unable to replicate the `Windows CI pipeline` run fail locally (fail log [here](https://dev.azure.com/onnxruntime/onnxruntime/_build/results?buildId=1382352&view=logs&j=2c350baf-0cef-5ef1-cbeb-e5cd88c89a11&t=e2851d88-2248-5227-ac0b-d999486a4bf5)). The test `LSTMTest.BackwardCompute` passes for me. Any pointer on why its behaving different?
> 2. The other failure in `Windows GPU CI Pipeline` (see [log](https://dev.azure.com/onnxruntime/onnxruntime/_build/results?buildId=1382353&view=logs&j=3032dfba-5baf-5872-0871-2e69cb7f4b6a&t=91e76d45-ff6f-5804-5ba0-eca18ac9ce48)) fails a test `onnxruntime_test_all` -> `ModelTests/ModelTest.Run/fp16_coreml_FNS_Candy_opset7_CPU`. But my local build doesn't contain the `ModelTests*` set at all. How do I build these tests?

The 1st failure is with DML EP on. Did you build with dml ep enabled?
For the 2nd issue, @snnn , is it possible that we can share the model?

r-devulap commented:
> The 1st failure is with DML EP on. Did you build with dml ep enabled? 

Nope, let me try building with DirectML. 

> For the 2nd issue, @snnn , is it possible that we can share the model?

You probably don't need to if we can just help figure out why this commit https://github.com/microsoft/onnxruntime/pull/20612/commits/5329f70d3b71f9261add1b1307f6c99364b8b624 isn't being applied to this model  `ModelTests/ModelTest.Run/fp16_coreml_FNS_Candy_opset7_CPU`.

snnn commented:
> For the 2nd issue, @snnn , is it possible that we can share the model?

Sorry, we cannot. We may consider removing these models from our build pipelines if there is no better way to handle them.

r-devulap commented:
1. Removed `if (provider_name == "cpu")` for  `fp16_coreml_FNS` test filter 
2. Added relative and absolute error tolerance for `LSTM.BackwardCompute` test. 

Hoping it will fix it, I still am not able to reproduce the failure locally though :/

snnn commented:
/azp run ONNX Runtime Web CI Pipeline,Linux CPU CI Pipeline,Linux GPU CI Pipeline,orttraining-amd-gpu-ci-pipeline,Android CI Pipeline,iOS CI Pipeline,ONNX Runtime React Native CI Pipeline

snnn commented:
/azp run Linux OpenVINO CI Pipeline,Linux QNN CI Pipeline,MacOS CI Pipeline,orttraining-amd-gpu-ci-pipeline,orttraining-linux-ci-pipeline,orttraining-linux-gpu-ci-pipeline,orttraining-ortmodule-distributed,onnxruntime-binary-size-checks-ci-pipeline,Big Models,Android CI Pipeline

snnn commented on orttraining/orttraining/test/training_ops/cpu/rnn/lstm_test.cc:
```diff
@@ -165,7 +165,10 @@ TEST(LSTMTest, BackwardCompute) {
-      {9.02288f, 9.77558f, 4.23378f, 4.6432f, 1.92046f, 2.09879f, 1.87627f, 2.06453f});
+      {9.02288f, 9.77558f, 4.23378f, 4.6432f, 1.92046f, 2.09879f, 1.87627f, 2.06453f},
+       false,
+       1e-02,
```
Please run clang-format on this file.

r-devulap replied:
Done.

r-devulap commented:
> You need sign the license/cla agreement to move on.

CLA shows up as signed now.

snnn commented:
/azp run Linux OpenVINO CI Pipeline,Linux QNN CI Pipeline,MacOS CI Pipeline,orttraining-amd-gpu-ci-pipeline,orttraining-linux-ci-pipeline,orttraining-linux-gpu-ci-pipeline,orttraining-ortmodule-distributed,onnxruntime-binary-size-checks-ci-pipeline,Big Models,Android CI Pipeline

snnn commented:
/azp run ONNX Runtime Web CI Pipeline,Linux CPU CI Pipeline,Linux GPU CI Pipeline,orttraining-amd-gpu-ci-pipeline,Android CI Pipeline,iOS CI Pipeline,ONNX Runtime React Native CI Pipeline

snnn commented:
/azp run Linux OpenVINO CI Pipeline,Linux QNN CI Pipeline,MacOS CI Pipeline,orttraining-amd-gpu-ci-pipeline,orttraining-linux-ci-pipeline,orttraining-linux-gpu-ci-pipeline,orttraining-ortmodule-distributed,onnxruntime-binary-size-checks-ci-pipeline,Big Models,Android CI Pipeline

snnn commented:
@yufenglee , please help review

snnn commented:
/azp run Linux Android Emulator QNN CI Pipeline, Linux CPU Minimal Build E2E CI Pipeline, Linux GPU TensorRT CI Pipeline, Windows ARM64 QNN CI Pipeline, Windows CPU CI Pipeline, Windows GPU CI Pipeline, Windows GPU TensorRT CI Pipeline, Windows x64 QNN CI Pipeline

yufenglee commented:
/azp run ONNX Runtime Web CI Pipeline,Linux CPU CI Pipeline,Linux GPU CI Pipeline,orttraining-amd-gpu-ci-pipeline,Android CI Pipeline,iOS CI Pipeline,ONNX Runtime React Native CI Pipeline

yufenglee commented:
/azp run Linux OpenVINO CI Pipeline,Linux QNN CI Pipeline,MacOS CI Pipeline,orttraining-amd-gpu-ci-pipeline,orttraining-linux-ci-pipeline,orttraining-linux-gpu-ci-pipeline,orttraining-ortmodule-distributed,onnxruntime-binary-size-checks-ci-pipeline,Big Models,Android CI Pipeline

yufenglee commented:
/azp run Linux Android Emulator QNN CI Pipeline, Linux CPU Minimal Build E2E CI Pipeline, Linux GPU TensorRT CI Pipeline, Windows ARM64 QNN CI Pipeline, Windows CPU CI Pipeline , Windows GPU CI Pipeline, Windows GPU TensorRT CI Pipeline, Windows x64 QNN CI Pipeline

yufenglee commented:
@r-devulap, thanks a lot for your contribution! Could you please also share an analysis on the accuracy loss?

r-devulap commented:
@yufenglee I calculated the worst case ULP error for `tanh` alone and that turns out to be 1523 ULP (relative error of 9.0803982019e-05) occurring at 4.9999938011e+00. Between `[-4.0, 4.0]` the max ULP error is 872 (relative error of 5.2006424519e-05). Is there anything else you were interested in?

r-devulap commented:
ping @yufenglee Just a friendly reminder :)

yufenglee commented:
> @yufenglee I calculated the worst case ULP error for `tanh` alone and that turns out to be 1523 ULP (relative error of 9.0803982019e-05) occurring at 4.9999938011e+00. Between `[-4.0, 4.0]` the max ULP error is 872 (relative error of 5.2006424519e-05). Is there anything else you were interested in?

1523 ULP is a little too large (e-04). I'm not sure if it leads to E2E difference for some models.
